keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,1015 @@
+  a ServiceIPRange associated. It also handles the special case for the defaults ServiceIPRanges
+- 1 new controller that participates in the IPAddress deletion, that guarantees that each Service IP
+  allocated has its corresponding IPAddress object, recreating it if necessary.
+
+[![](https://mermaid.ink/img/pako:eNp9Uj1rwzAQ_SvixpIMXU0plKQFL8Ek2aoOV-nsKkhWkOSASfzfK9tywG7oLeJ9nR5IVxBWEmQgNHq_VVg5NLxmcQaGHchdlKDryPWz0Y0P5PKC-eBUXT1QPPv8Wop58YFGaUW9mED7R24Lq5Vo2dMcH9szjd7uQblNvt3fC-7Q0KJZr8-o2ZK8eJPSkff_rdhTyez3KR6zDdOMKNVhHJ45sJfben17TeB-SRYJ1NoKDOQ5zIJDz0V4twxXVJObwrACQ86gkvEJh_4cwg8Z4tB7JZXY6NA7u2htzjIG36UK1kFWova0AmyCPbS1gCy4hiZT-gnJ1f0CaG6sgw)](https://mermaid.live/edit#pako:eNp9Uj1rwzAQ_SvixpIMXU0plKQFL8Ek2aoOV-nsKkhWkOSASfzfK9tywG7oLeJ9nR5IVxBWEmQgNHq_VVg5NLxmcQaGHchdlKDryPWz0Y0P5PKC-eBUXT1QPPv8Wop58YFGaUW9mED7R24Lq5Vo2dMcH9szjd7uQblNvt3fC-7Q0KJZr8-o2ZK8eJPSkff_rdhTyez3KR6zDdOMKNVhHJ45sJfben17TeB-SRYJ1NoKDOQ5zIJDz0V4twxXVJObwrACQ86gkvEJh_4cwg8Z4tB7JZXY6NA7u2htzjIG36UK1kFWova0AmyCPbS1gCy4hiZT-gnJ1f0CaG6sgw)
+
+#### The kube-apiserver bootstrap process and the service-cidr flags
+
+Currently, the Service CIDRs are configured independently in each kube-apiserver using flags. During
+the bootstrap process, the apiserver uses the first IP of each range to create the special
+"kubernetes.default" service. It also starts a reconcile loop, that synchronize the state of the
+bitmap with the assigned IPs to the Services.
+
+With current implementation, that each kube-apiserver can boot with different ranges configured.
+There is no conflict resolution, each apiserver keep writing and deleting others apiservers allocator bitmap
+and Services.
+
+In order to be completely backwards compatible, the bootstrap process will remain the same, the
+difference is that instead of creating a bitmap based on the flags, it will create new
+ServiceIPRange object from the flags (flags configuration removal is out of scope of this KEP) 
+with a special label `networking.kubernetes.io/service-ip-cidr-default` set to true.
+
+With the new implementation only the Service IP Range configuration will be consistent, but not the
+default IP Family, that will still depend on the flags configuration of each individual apiserver.
+However, users still can provide consistency by using a webhook or admission controller for
+defaulting the IPFamily or IPFamilyPolicy fields on Services (to be revisited before Beta).
+
+There can be multiple default ServiceIPRange objects, also for the same IP family, since
+apiservers with different configuration can work, the only condition is that the defined
+service IP ranges contain the `kubernetes.default` Service.
+
+The new bootstrap process will be:
+
+```
+read_flags
+if invalid flags
+  exit
+get ServiceCIDR objects
+if empty || no default
+   create default ServiceCIDR from flags
+   run
+else
+   if default matches flags
+      run
+   else if overlaps
+      log/event
+      run
+   else
+      log/event periodically
```

Comment:
So, to answer my own question above, you would NOT create N ranges, but instead complain that the default doesn;t match "my" flags?

This is racy, though - what happens when multiple API servers hit the "no default" condition - they will create their own range and then you have multiple different defaults.

I think it might be simpler and more obvious like:

```
at starup:
	read_flags
	if invalid flags
		exit
	run default-service-ip-range controller
	run apiserver

controller:
	watch all ServiceCIDR objects labelled from-flags
		ignore being-deleted ranges
	wait for first sync

controller on_event:
	if no default range matching exactly my flags
		log
		create a ServiceCIDR from my flags
			generateName: "default-" or "from-flags-"
			from-flags label: "true"
	else if multiple
		log
		if multiple ranges match exactly my flags
			sort them by creation time
			log
			delete all but one
		endif
	endif
	if kubernetes.default does not exist
		create it
```

This is a little different and means the `kubernetes.default` logic has to accept any assigned IP, even if it is not "my own".  Whichever apiserver gets there first wins.

What do you think?  Having this as a distinct controller seems nice and contained.

Reply:
> This is a little different and means the kubernetes.default logic has to accept any assigned IP, even if it is not "my own". Whichever apiserver gets there first wins.

yeah, but who keeps it consistent?

This topic is growing and I don't have still clear what do we want to offer, current implementation is super rigid, once something is created it keeps updating it in a loop. There is no conflict resolution and different configurations don't work, they just screw the cluster.

If we want to allow dynamic configuration of ranges, it seems legit that users can resize or replace the default range, but this opens a lot of new scenarios, super complicated ones when there are inconsistent configurations.

Maybe for alpha we don't allow different configurations for the flags

> What do you think? Having this as a distinct controller seems nice and contained.

yeah, right now is an infinity for loop :smile: , that sounds much better